pkg/resources/management.cattle.io/v3/project/validator.go

```diff
@@ -0,0 +1,220 @@
+	projectQuota := newProject.Spec.ResourceQuota
+	nsQuota := newProject.Spec.NamespaceDefaultResourceQuota
+	if projectQuota == nil && nsQuota == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	fieldErr, err := checkQuotaFields(projectQuota, nsQuota)
+	if err != nil {
+		return nil, fmt.Errorf("error checking project fields: %w", err)
+	}
+	if fieldErr != nil {
+		return admission.ResponseBadRequest(fieldErr.Error()), nil
+	}
+	fieldErr, err = a.checkQuotaValues(&nsQuota.Limit, &projectQuota.Limit, oldProject)
+	if err != nil {
+		return nil, fmt.Errorf("error checking quota values: %w", err)
+	}
+	if fieldErr != nil {
+		return admission.ResponseBadRequest(fieldErr.Error()), nil
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+func checkQuotaFields(projectQuota *v3.ProjectResourceQuota, nsQuota *v3.NamespaceResourceQuota) (*field.Error, error) {
+
+	if projectQuota == nil && nsQuota != nil {
+		return field.Required(projectSpecFieldPath.Child(projectQuotaField), fmt.Sprintf("required when %s is set", namespaceQuotaField)), nil
+	}
+	if projectQuota != nil && nsQuota == nil {
+		return field.Required(projectSpecFieldPath.Child(namespaceQuotaField), fmt.Sprintf("required when %s is set", projectQuotaField)), nil
+	}
+
+	projectQuotaLimitMap, err := convert.EncodeToMap(projectQuota.Limit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode project quota limit: %w", err)
+	}
+	nsQuotaLimitMap, err := convert.EncodeToMap(nsQuota.Limit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode namespace default quota limit: %w", err)
+	}
+	if len(projectQuotaLimitMap) != len(nsQuotaLimitMap) {
+		return field.Invalid(projectSpecFieldPath.Child(projectQuotaField), projectQuota, "resource quota and namespace default quota do not have the same resources defined"), nil
+	}
+	for k := range projectQuotaLimitMap {
+		if _, ok := nsQuotaLimitMap[k]; !ok {
+			return field.Invalid(projectSpecFieldPath.Child(namespaceQuotaField), nsQuota, fmt.Sprintf("missing namespace default for resource %s defined on %s", k, projectQuotaField)), nil
+		}
+	}
+	return nil, nil
+}
+
+func (a *admitter) checkQuotaValues(nsQuota, projectQuota *v3.ResourceQuotaLimit, oldProject *v3.Project) (*field.Error, error) {
+	// check quota on new project
+	fieldErr, err := namespaceQuotaFits(nsQuota, projectQuota)
+	if err != nil || fieldErr != nil {
+		return fieldErr, err
+	}
+
+	// if there is no old project or no quota on the old project, no further validation needed
+	if oldProject == nil || oldProject.Spec.ResourceQuota == nil {
+		return nil, nil
+	}
+
+	// check quota relative to used quota
+	fieldErr, err = usedQuotaFits(&oldProject.Spec.ResourceQuota.UsedLimit, projectQuota)
+	if err != nil || fieldErr != nil {
+		return fieldErr, err
+	}
+
+	// check if default quota is enough to set on namespaces
+	namespaces, err := a.namespaceCache.List(labels.Set(map[string]string{"field.cattle.io/projectId": oldProject.Name}).AsSelector())
```

Comment:
Oof, good catch, this would only work for the local cluster.

Adding a new field to the project, as well as adding controller code to keep it up to date, doesn't make much sense if it's just in the service of this one validation, which, as you pointed out, is flawed, so I'm thinking it may be worth reconsidering this check and breaking feature parity slightly.

It's not clear from the original commit what the purpose of the check is. As it stands, changing the default namespace quota on the project does not make any changes to existing namespaces. So I'm not sure it really matters what the default is. All that really matters is that the total used limit is within the project limit. The default only matters the next time you create a namespace.

So I'm actually inclined to go with option 3 and just drop this validation. What do you think?

Reply:
Removed this validation in a separate commit, can continue discussing it or if there is no problem will squash it

Reply:
I'll resolve the thread, we can remove this as per offline discussion.